File: SystemProgramming/FirstFollow.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char p[10][20];
char resultFirst[10][20];
int num;
char EPS='#';

int checkSmall(char a) {
    if (!(a>=65 && a <=90)) {
        return 1;
    }
    else if(a=='|'){
        return 0;
    }
     else {
        return 0;
    }
}

int checkEps(char a[]) {
    int i;
    for(i=0;i<strlen(a);i++){
        if(a[i]=='#')
            return 0;
    }
    return 1;
}

void attachToResult(char ch,char nt) {
    int i,flag=0;
    for(i=0;i<num;i++)
        if( p[i][0] == nt) 
            break;
    int index = i;     
    i=0;                   
    while(resultFirst[index][i] != 0) {
        if(resultFirst[index][i] == ch ) {
            flag=1;
            break;
        }            
        i++;
    }        
    if(flag ==0) {
        resultFirst[index][i]=ch;
        resultFirst[index][++i]='\0';
    }    
}

char * firstOf(char a[]) {    
    char nt,fch='\0'; 
    char * localresult;  
    char str[10]={""}; 
    int k=0;
    nt=a[0];
    int flag=0;
    int index = 3,i;
    EPS='\0';
    if(nt!=a[3]) {   
        while(a[index]!=0) {
            while(a[index]!='|') {
                int f = checkSmall(a[index]);
                if(f==1) {
                    //attachToResult(a[index],a[0]);                    
                    //printf("%c, ",a[index]);   
                    str[k++]=a[index];
                    strcat(localresult,str);
                    flag=1;                                                     
                    break;                    
                }
                else {
                    for(i=0;i<num;i++){
                        if( p[i][0] == a[index]) {
                                
                                //attachToResult(fch,a[0]);                              
                                strcat(str,firstOf(p[i]));
                                if(checkEps(p[i]))
                                    return '\0';
                                //attachToResult('#',a[0]);
                                fch = '#';                                                                
                                k = strlen(str)-1;
                                str[k++]=fch;
                                strcat(localresult,str);
                        }
                    }
                }
                index++;
            }
            if (flag==1)
            {
                while(a[index]!='|' && a[index]!=0) {
                    index++;
                }   
            }            
            flag=0;
            index++;
        }         
    } else {
        puts("Grammar is LEFT RECURSIVE\n");
    }
    return localresult;
}

int main() {
    char nt;
    char b;
    char first[20];
    int index=3,i;
    printf("Number of Production : \n");
    scanf("%d",&num);
    printf("Enter the grammar without space\n");
    for(i=0;i<num;i++){
        scanf("%s",p[i]);
    }
    for(i=0;i<num;i++){
        EPS='\0';
        printf("First(%s) ",p[i]);
        
        printf("%s \n",firstOf(p[i]));
        //puts("\n");
    }
}

File: SystemProgramming/FirstFollow.py
EPSILON = "#"
#code for printing epsilon
#print u"\u03B5"

def checkEpsilon(a,dict):
    if EPSILON in dict[a]:
        return True
    else:
        return False

def First(a,dict,result,ep):  
    tep=False    
    ep=False   
    flag =0
    if not(a.isupper()):
        result.add(a)
        return False
    if dict.get(a) == None :
        print "Incomplete Productions"
        return 
    p = dict[a].split("|")
    for x in p:
        if not(x[0]==a):
            for ch in x:  
                flag=0          
                if(not(ch.isupper())):
                    if ch != EPSILON:
                        result.add(ch) 
                        ep=False   
                        flag=1                
                        break                    
                else:
                    First(ch,dict,result,ep)
                    if not(checkEpsilon(ch,dict)):                        
                        ep=False
                        flag=1
                        break                                            
                    else:
                        if not(ep):
                            ep=True
                if(flag==1):
                    break
            if ep and not(tep):
                tep=ep
                
        else:
            print "LEFT RECURSIVE Grammar"
    if (checkEpsilon(a,dict)):
        tep=checkEpsilon(a,dict)
    # print result,tep
    return tep 



def Follow(a,dict,result,startSymbol):
    keys = dict.keys()    
    if (a in keys):
        if(a==startSymbol):
            result.add("$")
        for x in keys:
            rSide = dict[x]             
            for i in range(0,len(rSide)):
                if(a==rSide[i]):
                    # print rSide[i]           
                    if(i+1!=len(rSide)):  
                        if rSide[i+1] != '|' :
                            flag =0
                            tempResult = set()
                            ep = False 
                            for j in range(i+1,len(rSide)):
                                if(rSide[j] =='|'):
                                    break
                                if First(rSide[j],dict,result,ep):
                                    # print result
                                    flag=1
                                    continue                                    
                                else :
                                    # print result   
                                    flag = 0                                                                                       
                                    break
                            if flag==1 :
                                if(x!=a):                                
                                    Follow(x,dict,result,startSymbol)        
                                # print result
                        else:
                            if(x!=a):
                                Follow(x,dict,result,startSymbol)
                            # print result
                    else:
                        if(x!=a):
                                Follow(x,dict,result,startSymbol)
                        # print result                 
    else:
        print "Productions are incomplete"



def main():
    print "Enter number of production"
    n = input()    
    a={}
    startSymbol = ""
    result = set()
    for i in range(0,n):
        x = raw_input()
        y = x.strip().split("->")          
        if i==0:
            startSymbol = y[0]
        if a.get(y[0]) == None :
            if y[0].isupper() :
                a[y[0]]=y[1]        
            else:
                print "Not valid Production"
        else:
            sep = "|"
            temp = a[y[0]] 
            temp = temp,y[1]
            a[y[0]]=sep.join(temp)    
    ch = ""
    # print "Enter Terminal to find Follow"
    # ch = raw_input().strip().split()[0]       
    # result.clear()
    # Follow(ch,a,result,startSymbol)
    # print result  
    # result.clear() 

    while True and ch!='0' :
        print "Find First exit(0)"
        ch = raw_input().strip().split()[0]   
        if(ch=="0"):
            break
        ep = False 
        if First(ch,a,result,ep):
            result.add(EPSILON)
            print result 
        else :
            print result
        print "Find Follow exit(0)"
        ch = raw_input().strip().split()[0]   
        if(ch=="0"):
            break
        result.clear()
        Follow(ch,a,result,startSymbol)
        print result  
        result.clear()  


main()